File: Algo/Leetcode/037SolveSudoku.cpp
// dfs and trace back
class Solution {
public:
    bool isValid(vector<vector<char>>& board, int row, int col, char c) {
        int n = board.size();

        for(int i=0; i<n; i++){
            if(board[i][col]==c) return false;
        }

        for(int j=0; j<n; j++){
            if(board[row][j]==c) return false;
        }

        for(int i=0; i<3; i++){
            for(int j=0; j<3; j++){
                if(board[i+row/3*3][j+col/3*3]==c) return false;
            }
        }

        return true;
    }

    void solveSudoku(vector<vector<char>>& board) {
          dfs(board);
    }

    bool dfs(vector<vector<char>>& board){
        int n = board.size();
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(board[i][j] != '.') continue;
                for(int k=1; k<=9; k++){
                    char c = k+'0';
                    if(!isValid(board, i, j, c)) continue; // just check whether the intended input is correct
                    board[i][j] = c;
                    if(dfs(board)) return true;
                    board[i][j] = '.';
                }
                return false;
            }
        }
        return true;
    }
};

File: Algo/Leetcode/037SolveSudoku.py
class Solution(object):

    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        self.dfs(board)
        return board

    def dfs(self, board):
        n = len(board)
        m = len(board[0])

        for i in range(n):
            for j in range(m):
                if board[i][j] == '.':
                    for k in range(1, 10):
                        if self.checkValid(board, str(k), i, j, n, m):
                            board[i][j] = str(k)
                            res = self.dfs(board)
                            if res:
                                return True
                            board[i][j] = '.'
                    return False  # can not find any match this

        return True

    def checkValid(self, board, k, i, j, n, m):
        valid = True
        # check in col
        for t in range(n):
            if t != i:
                if board[t][j] == k:
                    valid = False
        if not valid:
            return False
        # check in row
        for t in range(m):
            if t != j:
                if board[i][t] == k:
                    valid = False
        if not valid:
            return False
        # check in small square
        start_i = i / 3
        start_j = j / 3
        for q in range(start_i * 3, start_i * 3 + 3):
            for w in range(start_j * 3, start_j * 3 + 3):
                if q != i and w != j:
                    if board[q][w] == k:
                        valid = False

        return valid
